### DISCUSSION TITLE
BLE (Bluetooth) BMS notify instead of polling?

### DISCUSSION THREAD

**Discussion by seidler2547** [Louisvdw/dbus-serialbattery#684] [Ideas]
Hi,

I have played around with my JKBMS over BLE and with a few small code changes I was able to remove the polling loop and rather publish new values to dbus as soon as a complete packet is decoded. Noticeable change is that JKBMS actually sends updates every 500--600ms, so twice as fast as the current polling interval.

I have not yet had the time to update to the latest version of either Venus OS or dbus-serialbattery, but if it's of interest I can see if I can make the time for that and make the changes there. But the changes are really easy, basically don't start the poller if it's a BLE battery, add a new_values_callback lambda variable to the battery class and call it there after self.decode().

Stefan

**Comment by mr-manuel** [comment]
@idstein would this also be something the the JBD BLE driver?

**Comment by idstein** [comment]
It is possible, but Venus OS should not process at a faster interval. We need async dbus for better efficiency before starting to increase the frequency. 
In general this feature is valid for all BMS solutions. The JBD BLE implementation irrigstes decode logic from RS485. So decreasing the poll interval should to the trick. gobject interval callbacks should also be deferred if the callback takes more than the given interval.

**Reply by seidler2547** [reply]
Yes I saw that JBD BLE always sends something, then start notify, waits until the values arrive and then stops notify. JKBMS just starts notify and new values keep arriving. I've been running it this way with sending new data to dbus as soon as it arrives for a month now and it works well. Maybe it's worth testing if the sending of the command is actually necessary for JBD or if it could also just keep on waiting for new data to arrive.

**Reply by idstein** [reply]
@seidler2547 correct. This was a lessons learned I had with JBD BMS and bad BLE connection. By re-starting notify (minimal overhead) for each request-reply, we reset on the JBD BMS BLE dongle the serial buffer to the main MCU. This allows to skip data from past communication or left-overs.

**Reply by mr-manuel** [reply]
@seidler2547 feel free to open a PR in the `dev` branch for this

**Reply by seidler2547** [reply]
https://github.com/Louisvdw/dbus-serialbattery/pull/689
